config: add `--when.commands` scope

bryceberger opened the pull request:
Closes #5217 

Adds a new config scope resolution `--when.commands`. This is a list of space-separated commands:
```toml
--when.command = ["file"]        # matches `jj file show`, `jj file list`, etc
--when.command = ["file show"]   # matches `jj file show`, but *NOT* `jj file list`
--when.command = ["file", "log"] # matches `jj file` *OR* `jj log` (or subcommand of either)
```

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

bryceberger commented:
Unsure how to update `config-schema.json` for this. The current entry for pager just says `"type": "string"`, which isn't true.

Also unsure if / where there are tests for pagers. I was unable to find any on a cursory glance.

yuja commented:
> ```toml
> [ui]
> pager.status = ":none"
> pager.log = { command = ["less", "-FRX"], env = { LESSCHARSET = "utf-8" } }
> pager.diff = ["delta", "..."]
> pager.default = "..."
> ```

I think it's better to not overuse ":<type>" syntax. How about adding a separate table for per-table pager configuration?

```toml
[ui]
paginate = "auto"  # default
pager = "..."  # default

[<table-for-per-command-pager-options>]
status.paginate = "never"
diff.pager = ["delta", "..."]

# or generic per-command table like this?
[commands.status]
ui.paginate = "never"

# or extend the conditional table?
[[--scope]]
--when.commands = ["status"]
ui.paginate = "never"
```

> The full name of the highest level subcommand is used, i.e. `jj operation diff` uses `ui.pager.operation`.

I think "file show" and "file list" are quite different, and the user might want to enable pager only for one of them.

bryceberger commented:
I'm intrigued by extending the `[[--scope]]` rules.

The rule for `--when.repositories = ["a", "b"]` is to match `"a" | "b"` --- would have to put some thought into how to match `operation`, `operation diff`, and `diff` separately. Maybe `"operation.diff"`?

Currently, `jj_lib::config_resolver::resolve()` throws away the layer if its `--when` doesn't match. Thinking of changing `ScopeCondition::matches()` to return `Option<bool>`, with `None` meaning "not enough information to resolve", i.e. resolving `--when.repository` during `cli_util::run_internal` but waiting until `commands::run_command` to resolve `--when.commands`.

yuja commented:
> The rule for `--when.repositories = ["a", "b"]` is to match `"a" | "b"` --- would have to put some thought into how to match `operation`, `operation diff`, and `diff` separately. Maybe `"operation.diff"`?

or `"operation diff"`? The `[colors]` table uses space-separated string.

> Currently, `jj_lib::config_resolver::resolve()` throws away the layer if its `--when` doesn't match. Thinking of changing `ScopeCondition::matches()` to return `Option<bool>`, with `None` meaning "not enough information to resolve", i.e. resolving `--when.repository` during `cli_util::run_internal` but waiting until `commands::run_command` to resolve `--when.commands`.

`run_command()` calls `resolve()` multiple times, so I think it's okay to omit `--when.commands` conditionals until the command name becomes available. That's how `--repository` path is resolved.

ilyagr commented:
Just saw this, nice! 🎉 

FYI, I'm planning to add config for the built-in pager. My original thought was to simply create a `ui.builtin-pager` table. I'm not sure how it'll look if we make it configurable per-command, but we could figure it out later.

yuja commented:
> simply create a `ui.builtin-pager` table.

nit: might be better to include "streampager" in the key (e.g. `stream-pager`, `builtin-stream-pager`, `builtin-pager-sp`) so we can freely switch the underlying implementations in future.

> I'm not sure how it'll look if we make it configurable per-command,

The current idea is to extend the scoped table, so any parameters (except for `[aliases]`) can be set per-command.

ilyagr commented:
> > simply create a `ui.builtin-pager` table.
> 
> nit: might be better to include "streampager" in the key (e.g. `stream-pager`, `builtin-stream-pager`, `builtin-pager-sp`) so we can freely switch the underlying implementations in future.

I went back and forth on this. I went into the direction of making the options more generic, see https://github.com/jj-vcs/jj/pull/5415 . Your suggestion might still be better, I'm not sure. If we go with that, perhaps `:builtin` should be an alias for `:builtin-sp`.

> > I'm not sure how it'll look if we make it configurable per-command,
> 
> The current idea is to extend the scoped table, so any parameters (except for `[aliases]`) can be set per-command.

That sounds great!

bryceberger commented:
Completely changed the implementation, as reflected in the title. Now implements `--when.commands = [...]`.

Also changed how config resolution works --- a condition can either be "resolved true", "resolved false", or "unknown". If any field is "unknown", the entire expression is unknown, and kept around. If any field is "resolved false", the entire expression is false, and discarded. Otherwise, added to current config.

This means that if a config layer still contains a `--when` key after resolution, it was resolved to "unknown". These layers are skipped in `config::get_merged_item()`.

yuja commented:
> Also changed how config resolution works --- a condition can either be "resolved true", "resolved false", or "unknown". If any field is "unknown", the entire expression is unknown, and kept around. If any field is "resolved false", the entire expression is false, and discarded. Otherwise, added to current config.

Do we really need this? As I said, config layers are resolved multiple times in `run_internal()`, and the last resolution wins. If conditional variables have `--when.commands`, I don't think they should be disabled until the command name becomes available.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -3701,6 +3701,23 @@ impl CliRunner {
+                ui.reset(&config)?;
+                Some(subcommand_combined)
+            } else {
+                None
```
nit: perhaps, actual command name can be set to `""` if no subcommands exist. `None` means the command name is unknown, whereas the command is known to be unspecified here.

yuja commented on cli/src/config.rs:
```diff
@@ -395,9 +395,18 @@ impl ConfigEnv {
+        self.resolve_config_with_command(config, None)
+    }
+
+    pub fn resolve_config_with_command(
```
set the command name in `ConfigEnv`, and do `resolve_config()`. `config_env` can be reused later in order to resolve settings for cloned/inited workspace.

yuja commented on lib/src/config_resolver.rs:
```diff
@@ -98,11 +117,32 @@ fn expand_home(path: &Path, home_dir: Option<&Path>) -> Result<Option<PathBuf>,
+        // empty condition never matches
+        (Some([]), _) => Some(false),
+        (Some(candidates), Some(actual)) => {
+            let actual = actual.split(' ');
+            let any_match = candidates.iter().any(|candidate| {
+                let candidate = candidate.split(' ');
+                candidate.zip_longest(actual.clone()).all(|e| match e {
```
Maybe `.strip_prefix()` and check the first character?

bryceberger replied:
Ah, that is cleaner. Not sure why you would need to check the first character --- `"file show".strip_prefix("file") == Some(" show")`, `"file show".strip_prefix("file show") == Some("")`.

bryceberger replied:
Oh, so `jj abcdef` doesn't accidentally match `--when.commands = ["abc"]`.

bryceberger commented:
> Do we really need this? As I said, config layers are resolved multiple times in `run_internal()`, and the last resolution wins. If conditional variables have `--when.commands`, I ~don't~ think they should be disabled until the command name becomes available.

... oooh, the `RawConfig` is kept unchanged through resolution. I had it in my mind that `pop_scope_condition()` would destructively remove the table, didn't realize it was clone-on-write.

yuja commented on lib/src/config_resolver.rs:
```diff
@@ -106,6 +119,18 @@ fn matches_path_prefix(candidates: Option<&[PathBuf]>, actual: Option<&Path>) ->
 
+fn matches_command(candidates: Option<&[String]>, actual: Option<&str>) -> bool {
+    match (candidates, actual) {
+        // empty condition never matches
+        (Some([]), _) => false,
```
nit: I don't think optimization is needed for empty condition.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -3701,6 +3701,25 @@ impl CliRunner {
         }
 
+        config_env.set_command(
+            if let Some((subcommand, mut subcommand_matches)) = matches.subcommand() {
+                let mut subcommand_combined = String::from(subcommand);
+                while let Some((subcommand, new_matches)) = subcommand_matches.subcommand() {
+                    subcommand_combined.push(' ');
+                    subcommand_combined.push_str(subcommand);
+                    subcommand_matches = new_matches;
+                }
```
nit: I would extract this part to a function that returns string command name.

yuja commented on cli/src/cli_util.rs:
```diff
@@ -3701,6 +3701,25 @@ impl CliRunner {
+                    subcommand_combined.push_str(subcommand);
+                    subcommand_matches = new_matches;
+                }
+                config = config_env.resolve_config(&raw_config)?;
+                migrate_config(&mut config)?;
+                ui.reset(&config)?;
```
nit: deduplicate this. Perhaps, `ui.reset()` can be omitted as we do that later.

yuja commented on cli/src/config.rs:
```diff
@@ -281,9 +282,14 @@ impl ConfigEnv {
         })
     }
 
+    pub fn set_command(&mut self, command: Option<String>) {
```
nits: 
* `set_command_name()` for clarity? (Perhaps, context variable can also be renamed.)
* just take `String`? It's unlikely that we need to clear the command name.

yuja commented on docs/config.md:
````diff
@@ -1371,6 +1371,9 @@ You can conditionally enable config variables by using `--when` and
+are specified, the intersection is used.
+
 ```toml
 [user]
````
nit: can you add pager example here?

martinvonz commented:
I just noticed that [the documentation](https://jj-vcs.github.io/jj/latest/config/#conditional-variables) has not been updated. Are you able to fix that, @bryceberger?

ilyagr commented:
There [prerelease docs](https://jj-vcs.github.io/jj/prerelease/config/#conditional-variables) have some information. I'm not sure whether something is missing in there as well.

martinvonz commented:
No, that looks good. Sorry, I don't know why I went to the 0.25 docs.
